### DISCUSSION TITLE
Automating test templates, bolierplate, skeletons

### DISCUSSION THREAD

**Discussion by bobturneruk** [pykale/pykale#71] [Testing] [answered]
@haipinglu - is this quick and dirty example of using the language parser to pull out function names any use? One could maybe extend this to generate template text files as the basis for writing tests.

```
import ast

# I've hard coded `print.py` for the purposes of this gist
with open('print.py', 'r') as reader:
    # Read & print the entire file
    parsed = ast.parse(reader.read())
    # Parse python language, filter out function definitions
    function_definitions = [n for n in ast.walk(parsed) if type(n) == ast.FunctionDef]
    # print function names
    print([definition.name for definition in function_definitions])
```

```
> python .\find_funs.py
['tprint', 'pprint', 'pprint_without_newline']
```

Thoughts from @drj11 / @willfurnass appreciated...

**Comment by haipinglu** [comment]
That's helpful. Thanks!
@bobturneruk are these relevant?
- https://pypi.org/project/PyTestStub/
- https://github.com/laffra/auger
- https://pypi.org/project/pynguin/
- https://pypi.org/project/pythoscope/

**Comment by bobturneruk** [comment, marked as answer]
- https://pypi.org/project/PyTestStub/
> This uses `unittest`, not `pytest`, so probably not great for us.

- https://github.com/laffra/auger
> Again uses `unittest` which will work but creates more boilerplate than we need.

- https://pypi.org/project/pynguin/
> This looks interesting and might have saved me some time! It is `pytest`-based. However, I'm not sure how it will handle non-base data types e.g. from `numpy` or `pandas`. Might be worth a try.

- https://pypi.org/project/pythoscope/
> It's about a decade since the last release of this and I can't find much information. I suggest it won't be that useful.

**Comment by haipinglu** [comment]
@bobturneruk I've been looking for a compact self-contained (and printable) tutorial doc on pytest and found this [pytest tutorial](https://www.tutorialspoint.com/pytest/pytest_tutorial.pdf) to be good. It covers quite some essentials and takes less than one hour to go through. How do you think? If yes, we could add it to the readme and share it with the team.

**Comment by bobturneruk** [comment]
Looks great - I've learned a couple of things (about `conftest.py` and parallel tests) from it!
